File: solutions/Golang/binary_tree_level_order_traversal_bfs.go
package main

type TreeNode struct { Val int; Left, Right *TreeNode }

func LevelOrder(root *TreeNode) [][]int {
    var result [][]int
    if root == nil { return result }
    queue := []*TreeNode{root}
    for len(queue) > 0 {
        sz := len(queue)
        level := make([]int, 0)
        for i := 0; i < sz; i++ {
            n := queue[0]
            queue = queue[1:]
            level = append(level, n.Val)
            if n.Left != nil { queue = append(queue, n.Left) }
            if n.Right != nil { queue = append(queue, n.Right) }
        }
        result = append(result, level)
    }
    return result
}

File: solutions/Rust/binary_tree_level_order_traversal_bfs.rs
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    let mut result = Vec::new();
    if root.is_none() { return result; }
    let mut queue = VecDeque::new();
    queue.push_back(root.unwrap());
    while !queue.is_empty() {
        let sz = queue.len();
        let mut level = Vec::new();
        for _ in 0..sz {
            let n = queue.pop_front().unwrap();
            let n_ref = n.borrow_mut();
            level.push(n_ref.val);
            if let Some(l) = n_ref.left.clone() { queue.push_back(l); }
            if let Some(r) = n_ref.right.clone() { queue.push_back(r); }
        }
        result.push(level);
    }
    result
}
